### PR TITLE
Fixes #17653 - Allow all parameters api to unhide values

### PR DESCRIPTION
**Pull request by orrabin**

### CODE REVIEW

File: app/models/lookup_keys/puppetclass_lookup_key.rb [hunk 1]
```diff
@@ -20,6 +20,10 @@ class PuppetclassLookupKey < LookupKey
 
   scope :smart_class_parameters, -> { joins(:environment_classes).readonly(false) }
 
+  def can_edit?
```
Comment (tbrisker) [inline app/models/lookup_keys/puppetclass_lookup_key.rb]: This name is a bit confusing - the LK isn't doing the editing. Perhaps change it to `editable_by_user?` or something similar?

File: app/models/lookup_keys/puppetclass_lookup_key.rb [hunk 2]
```diff
@@ -20,6 +20,10 @@ class PuppetclassLookupKey < LookupKey
...
+    User.current.allowed_to?(:edit_external_parameters)
```
Comment (tbrisker) [inline app/models/lookup_keys/puppetclass_lookup_key.rb]: This checks if the user has that permission, but what if it is only limited to certain parameters?

File: app/views/api/v2/common_parameters/base.json.rabl [hunk 3]
```diff
@@ -1,3 +1,10 @@
 object @common_parameter
 
-attributes :id, :name, :value
+attributes :id, :name
+
+if params["show_hidden"] && User.current.allowed_to?(:edit_params)
```
Comment (tbrisker) [inline app/views/api/v2/common_parameters/base.json.rabl]: this is quite repetitive, perhaps extract to a concern that is used by both params and lks?

File: test/unit/rabl_test.rb [hunk 4]
```diff
@@ -3,14 +3,14 @@
 
 class RablTest < ActiveSupport::TestCase
   test 'render of single template' do
-    @domain = FactoryGirl.build(:domain)
```
Comment (tbrisker) [inline test/unit/rabl_test.rb:6]: Are these changes related or added by mistake to this commit?

**Review by tbrisker — CHANGES_REQUESTED** [review]
Nice work on the test coverage! just a few fairly minor inline comments.

File: test/unit/rabl_test.rb (hunk 4 above)
Comment (orrabin) [reply test/unit/rabl_test.rb:6]: Because it's only testing the rendering I didn't want to make changes to fit the new permissions so I changed it to `medium` which doesn't have parameters.

File: app/views/api/v2/common_parameters/base.json.rabl (hunk 3 above)
Comment (orrabin) [reply app/views/api/v2/common_parameters/base.json.rabl]: The concern would need to know when to use view and default_view and which permission it uses.

**Comment by orrabin** [comment]
@tbrisker thanks for your comments, I addressed most of them, resolving the duplication would need more `if`s then it's worth in my opinion  but I don't mind changing it.

File: app/models/lookup_keys/puppetclass_lookup_key.rb [hunk 5]
```diff
@@ -20,6 +20,10 @@ class PuppetclassLookupKey < LookupKey
 
   scope :smart_class_parameters, -> { joins(:environment_classes).readonly(false) }
 
+  def editable_by_user?
+    PuppetclassLookupKey.authorized(:edit_external_parameters).present? && PuppetclassLookupKey.authorized(:edit_external_parameters).find(self.id)
```
Comment (tbrisker) [inline app/models/lookup_keys/puppetclass_lookup_key.rb]: This will cause two queries on the DB, better to do something like:
```ruby
PuppetclassLookupKey.authorized(:edit_external_parameters).where(:id => id).exists?
```
which will generate a much simpler query.

File: app/models/lookup_value.rb [hunk 6]
```diff
@@ -1,10 +1,11 @@
 class LookupValue < ActiveRecord::Base
   include Authorizable
   include PuppetLookupValueExtensions
+  include HiddenValue
 
   validates_lengths_from_database
   audited :associated_with => :lookup_key
-  delegate :hidden_value?, :hidden_value, :to => :lookup_key, :allow_nil => true
+  delegate :hidden_value?, :editable_by_user?, :to => :lookup_key, :allow_nil => true
```
Comment (tbrisker) [inline app/models/lookup_value.rb:8]: no need for the `allow_nil` here, these are not attributes now

File: app/views/api/v2/common_parameters/base.json.rabl [hunk 7]
```diff
@@ -1,3 +1,11 @@
 object @common_parameter
 
-attributes :id, :name, :value
+attributes :id, :name
+
+authorization = Parameter.authorized(:edit_params)
+if params["show_hidden"] && @common_parameter.present? && authorization.present? && authorization.find(@common_parameter.id)
```
Comment (tbrisker) [inline app/views/api/v2/common_parameters/base.json.rabl]: how about making an `editable_by_user?` method on parameter model as well?

**Review by tbrisker — CHANGES_REQUESTED** [review]
I still think that removing the duplication in the api is worth it, and if you create a `editable_by_user` method for parameters as well it should reduce the need for ifs (as you would just call it on whatever object)

File: app/models/lookup_value.rb (hunk 6 above)
Comment (orrabin) [reply app/models/lookup_value.rb:8]: if the lookup_key is nil and I remove the allow_nil there will be an error.

File: app/views/api/v2/common/show_hidden.json.rabl [hunk 8]
```diff
@@ -0,0 +1,7 @@
+object @parameter
+
+if params["show_hidden"] == "true" && @object.present? && @object.editable_by_user?
+  attribute locals[:value].to_sym
```
Comment (tbrisker) [inline app/views/api/v2/common/show_hidden.json.rabl]: why not pass the field name as a symbol instead of passing as a string and calling `to_sym`?

File: app/views/api/v2/override_values/base.json.rabl [hunk 9]
```diff
@@ -3,3 +3,7 @@ object @override_value
 attributes :id, :match, :value, :omit
 # compatibility
 attribute :omit => :use_puppet_default
+
+node do
+  partial("api/v2/common/show_hidden", :locals => {:value => 'value'}, :object => @object)
+end
```
Comment (tbrisker) [inline app/views/api/v2/override_values/base.json.rabl]: nitpick - missing newline at EOF

File: app/views/api/v2/common/show_hidden.json.rabl [hunk 10]
```diff
@@ -0,0 +1,7 @@
+object @parameter
+
+if params["show_hidden"] == "true" && @object.present? && @object.editable_by_user?
```
Comment (tbrisker) [inline app/views/api/v2/common/show_hidden.json.rabl]: is the `@object.present?` check still needed? wouldn't a missing object cause a failure anyways when attempting to access the attribute?

**Review by tbrisker — CHANGES_REQUESTED** [review]
Almost done, just a few more minor comments inline

**Comment by tbrisker** [comment]
Thanks @orrabin !
